tests/baselines/reference/dependentReturnType2.errors.txt

```diff
@@ -0,0 +1,314 @@
+    // File: Rocket.Chat/apps/meteor/app/settings/lib/settings.ts
+    
+    /**
+     * @typedef {Record<any, any>} MyObj
+     */
+    
+    
+    /**
+     * @typedef {MyObj} SettingValue
+     */
+     
+    /**
+     * @template {SettingValue} T
+     * @typedef {Object} SettingComposedValue
+     * @property {string} key
+     * @property {SettingValue} value
+     */
+    
+    /**
+     * @callback SettingCallback
+     * @param {string} key
+     * @param {SettingValue} value
+     * @param {boolean} [initialLoad]
+     * @returns {void}
+     */
+    
+    /** @type {{ settings: { [s: string]: any } }} */
+    const Meteor = /** @type {any} */ (undefined);
+    /** @type {{ isRegExp(x: unknown): x is RegExp; }} */
+    const _ = /** @type {any} */ (undefined);
+    
+    /**
+     * @param {RegExp} x
+     * @returns {void}
+     */
+    function takesRegExp(x) {
+        return /** @type {any} */ undefined;
+    }
+    /**
+     * @param {string} x
+     * @returns {void}
+     */
+    function takesString(x) {
+        return /** @type {any} */ undefined;
+    }
+    
+    /**
+     * @class NewSettingsBase
+     */
+    class NewSettingsBase {
+        /**
+         * @template {SettingCallback | undefined} C
+         * @template {string | RegExp} I
+         * @template {SettingValue} T
+         * @param {I} _id
+         * @param {C} [callback]
+         * @returns {HelperCond<C, SettingCallback, void, undefined, HelperCond<I, string, T | undefined, RegExp, SettingComposedValue<T>[]>>}
+         */
+        newGet(_id, callback) {
+            if (callback !== undefined) {
+                if (!Meteor.settings) {
+                    return; // Ok
+                }
+                if (_id === '*') {
+                    return Object.keys(Meteor.settings).forEach((key) => {
+                        const value = Meteor.settings[key];
+                        callback(key, value);
+                    });
+                }
+                if (_.isRegExp(_id) && Meteor.settings) {
+                    return Object.keys(Meteor.settings).forEach((key) => {
+                        if (!_id.test(key)) {
+                            return;
+                        }
+                        const value = Meteor.settings[key];
+                        callback(key, value);
+                    });
+                }
+    
+                if (typeof _id === 'string') {
+                    const value = Meteor.settings[_id];
+                    if (value != null) {
+                        callback(_id, Meteor.settings[_id]);
+                    }
+                    return; // Ok
+                }
+    
+                return; // Ok, needed for exhaustiveness check
+            }
+    
+            if (!Meteor.settings) {
+                return undefined; // Error
+                ~~~~~~
+!!! error TS2322: Type 'undefined' is not assignable to type 'HelperCond<I, string, T | undefined, RegExp, SettingComposedValue<T>[]>'.

```

Comment:
It's too bad we don't give any elaboration here.

Reply:
Yes, I don't really know how we could provide more feedback when we can't narrow a return type. It's already not 100% clear whether the user is trying to narrow the return type or not, and even when we think they are, I can't think of a good mechanism to explain why we didn't narrow.